src/v/resource_mgmt/cpu_profiler.h

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/property.h"
+#include "seastarx.h"
+
+#include <seastar/core/internal/cpu_profiler.hh>
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/sharded.hh>
+#include <seastar/core/smp.hh>
+#include <seastar/core/timer.hh>
+#include <seastar/util/backtrace.hh>
+
+#include <absl/container/node_hash_map.h>
+
+#include <chrono>
+#include <cstdint>
+#include <deque>
+
+namespace resources {
+
+/**
+ * The `cpu_profiler` service polls traces from seastar at fixed intervals.
+ * It then aggregates the traces and provides methods for other services
+ * to view them.
+ */
+class cpu_profiler : public ss::peering_sharded_service<cpu_profiler> {
+    // The number of results buffers to retain. Each of these buffers
+    // are `ss::max_number_of_traces` * 1568bytes in size (196KiB).
+    // Each buffer has samples from a window of
+    // `_sample_rate` * `ss::max_number_of_traces` milliseconds.
+    //
+    // At 10 results buffers with a 100ms sample rate this service
+    // retains samples representative of the last 2 minutes of CPU
+    // activity.
+    static constexpr size_t number_of_results_buffers{10};
+
+public:
+    struct cpu_profiler_result {
+        ss::sstring user_backtrace;
+        size_t occurrences;
+
+        cpu_profiler_result(ss::sstring ub, size_t o)
```

Comment:
nit: with a struct with public members like this I think the ctor serves nor purpose, you can create it in the same way using aggregate initialization

Reply:
For some reason `std::vector::emplace_back` wasn't working without this constructor. I'll try again though without it.

Reply:
Oh, it's a bug in clang/libc++. They don't accept aggregate initialization in some places they should, like `emplace_back`.  Very annoying:

https://stackoverflow.com/questions/13812703/c11-emplace-back-on-vectorstruct#comment126745383_61205386

I think rather than adding the ctor though you can just use emplace/push_back with a temporary: cpu_profiler_result{...} at the emplace_back call site.

Reply:
This is probably _slightly_ slower but errs on the side of clean code.